# Question 1

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

# Question 2

Questions 2.1

In [ ]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
villagers_df = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = villagers_df.shape

# Output the result
print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

# Optionally, you can also print the column names
print("\nColumn names:")
print(villagers_df.columns)

Number of rows: 391
Number of columns: 11

Column names:
Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')


Question 2.2

The term 'observation' essentially refers to the rows, or the individual, unique data entries from the dataset.  In this case, this refers to the individual characters, which there are 391 of.

The term 'variable' is referring to the differentiating factors of all the villagers.  These are the columns in the dataset, referring to things like name, gender, birthday, etc.

# Question 3

In [ ]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
villagers_df = pd.read_csv(url)

# Generate summary statistics for numeric columns
summary_statistics = villagers_df.describe()

# Get counts for categorical columns
species_counts = villagers_df['species'].value_counts()
personality_counts = villagers_df['personality'].value_counts()
gender_counts = villagers_df['gender'].value_counts()

# Output results
print("Summary Statistics:")
print(summary_statistics)

print("\nSpecies Counts:")
print(species_counts)

print("\nPersonality Counts:")
print(personality_counts)

print("\nGender Counts:")
print(gender_counts)



Summary Statistics:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

Species Counts:
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

Personality Counts:
personality
lazy      60
normal    59
cranky    55
snooty    55
jock      55
peppy     49
smug      34
uchi      24
Name: count, dtype: int64

Gender Counts:
gender
male      204
female 

# Question 4

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

# Display the shape of the dataset
print("Dataset Shape (rows, columns):", df.shape)

# Display the statistical summary of numeric columns
print("\nSummary Statistics (Numeric Columns Only):")
print(df.describe())

# Explanation of differences
print("\nExplanation:")
print("1. df.shape shows the total number of rows and columns.")
print("2. df.describe() only analyzes numeric columns, so non-numeric columns are not shown (like 'Sex', 'Name').")
print("3. The 'count' in df.describe() reflects the number of non-missing values for each numeric column.")


Dataset Shape (rows, columns): (891, 15)

Summary Statistics (Numeric Columns Only):
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Explanation:
1. df.shape shows the total number of rows and columns.
2. df.describe() only analyzes numeric columns, so non-numeric columns are not shown (like 'Sex', 'Name').
3. The 'count' in df.describe() refl

So the difference between df.shape and df.describe essentially lies between the differences in what they are analyzing.  The function df.shape analyzes both numeric variables (think age) and non numeric variables (think gender).  On the other hand, df.describe() solely looks at the numeric variables and shows calculations such as mean, standard deviation, etc.  

# Question 5

Essentially, methods versus attributes are different due to the differences in calculations.  Methods needs parentheses because they perform a calculation, such as finding the mean of age for example.  While attributes don't perform "calculations", rather they state things that are sort of "just true" about the dataset, such as the number of columns and rows.

#Question 6

Here are some quick definitions of all of the summary statistics provided by df.describe()

Count: provides the number of valid entries in each column.  Essentially, how many 'non-empty' values each column has.

Mean: an average of all of the data points in that specific column.  This is found by summing all of the individual data points, and then dividing it by the total number of valid data points (like the number in the 'count' statistic) in that column.

Minimun (min): The smallest value in that column.

25%: The first quartile, it is the point where 25% of the data lies below this value.

50%: The second quartile, the point where 50% of the data lies below that value, also referred to as the median.

75%: The third quartile, the point where 75% of the data lies below this value.

Maximum (max): The largest value in that column.

# Question 7

1. Using df.dropna() would be beneficial in cases when you don't have a large amount of data missing, but relatively smaller amounts across rows.
Take the Titanic dataset below.

In [1]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Check for missing values in each column
missing_data = df.isnull().sum()

# Print the total number of missing values and which columns contain missing data
print(missing_data)


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


Only 2 passengers have missing data in the 'embark_town' column, and it would be unreasonable to delete this entire column for that reason, therefore it is a better idea to use df.dropna() for this case.

2. Using the missing values table from questionn 1, we can see how the 'deck' column has 688 missing values.  This is an overwhelming majority of the data which does not have values here, and using df.dropna() to remove every one of these rows would be completely unreasonable as a large amount of the other columns of data would be removed.  In this case, using del df['col'] would be a better choice.


3. Using del df['col'] before df.dropna() is a better idea because let's say a majority of the data in a column is missing (say 70%+), if you used df.dropna() first on these rows, that would mean that a huge amount of usable data will be lost.  As opposed to using del df['col'] first, in which case this entire column will be delete, and then you can individually take out other rows using df.dropna() where you can retain as much usable data as possible.

4. In order to most efficiently remove data from the dataset, I will remove the 'deck' column first, and then remove the rows of missing data from 'age', 'embarked', and 'embark_town'.

In [2]:
# Load the Titanic dataset again to ensure a fresh start
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Check initial shape of the dataset
print("Initial shape: ", df.shape)

# Step 1: Delete the 'deck' column, which has too many missing values
del df['deck']

# Report shape after deleting 'deck'
print("After deleting 'deck': ", df.shape)

# Step 2: Drop rows where essential columns have missing values (age, embarked and embark_town)
df_clean = df.dropna(subset=['age', 'embarked', 'embark_town'])

# Report shape after dropping missing values
print("After dropping rows with missing values: ", df_clean.shape)


Initial shape:  (891, 15)
After deleting 'deck':  (891, 14)
After dropping rows with missing values:  (712, 14)


Justification: As you can see, we were able to still keep a majority of the data.  Keeping 712 out of 891 rows and 14 out of 15 columns, if we used a different approach (say removing the rows of missing data for 'deck' first) we would end up with a much smaller set of data.

#Question 8

In [4]:


# Group by 'class' and summarize statistics for 'fare'
grouped_summary = df.groupby("class")["fare"].describe()
print(grouped_summary)


        count       mean        std  min       25%      50%   75%       max
class                                                                      
First   216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
Second  184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
Third   491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500


1. As you can see from code above, I chose to compare class to fare.  What we can see is that putting class as 'col1' results in class being on the rows, and the different 'fare' values are analysed against the class using the .describe() function's different numerical stats.  What this is essentially doing is analysing all of the different statistics of each class and presenting them so we can compare, for example, we can see how in the 'mean' category, first class had paid a lot more than second or third, which of course makes sense.  However this provides us with the ability to make more niche, less obvious conclusions.

2. The key difference between groupby() and describe() is that describe() provides the overall summary stats for the entire dataset.  Specifically for all non-missing values.  On the other hand, groupby() specifically breaks down the stats in the 'col2' into the different groups of 'col1'. Using the previous part (1) as an example, the fare is split into the categories 'First', 'Second' and 'Third', whereas if we used just describe(), it would provide stats for  the entire dataset, not broken down into specific categories.

3. IN ORDER TO NOT LOSE THE CODE I HAVE ALREADY RAN, I RAN THE ERRORS IN A NEW NOTEBOOK AND PASTED THE LINK FOR THE NOTEBOOK ON MY REPO HERE.

https://github.com/nehan-p/sta130/blob/main/STA130HW1ErrorQuestion.ipynb


# Chatbot Summary

Here is the link for my chatbot conversation and a summary of the chat.

NOTE - I was not able to download the direct link for my conversation through ChatGPT since an error kept occurring, so I downloaded the pdf and uploaded it to my github, which the link for is here.

https://github.com/nehan-p/sta130/blob/main/Dataset%20Columns%20Summary.pdf


Dataset Exploration:

You were working with the Titanic dataset from the URL https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv.
We explored missing data using df.isnull().sum() to check how much data was missing and where.
We discussed the differences between df.shape and df.describe(), explaining why they give different information due to non-numeric columns and missing values.
Explanation of Summary Statistics:

I provided definitions for the various summary statistics generated by df.describe(), such as 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max'.
Handling Missing Data:

We explored two approaches to dealing with missing data: df.dropna() and del df['column'].
You learned about use cases for when each method might be preferred and how they can be applied to optimize the use of non-missing data.
GroupBy Operation:

You experimented with using df.groupby("class")["alive"].describe() to get group-level statistics and learned why not all describe statistics are shown (for categorical data only certain statistics like count, unique, top, and freq are relevant).
Error Debugging:

We walked through some common coding errors like NameError: name 'pd' is not defined, fixing typos, missing parentheses, and handling column name errors.
I explained how to simulate and fix these errors, including restarting the kernel and re-importing libraries in Google Colab.
GitHub Upload:

You asked how to upload your Jupyter notebook to GitHub. I guided you through saving the notebook from Google Colab, uploading it to GitHub via the web interface, and then submitting the link to your Quercus assignment.
Final Steps:

You requested help with inserting screenshots into your Google Colab notebook and I explained how to do this.
I provided instructions for uploading the notebook to GitHub and submitting it for your assignment.
